Default to opaque white color in more `CanvasItem::draw_*` methods

kleonc opened the pull request:
Makes the `color` parameter default to opaque white (`Color(1, 1, 1, 1)`) for these methods:
- `CanvasItem::draw_dashed_line`,
- `CanvasItem::draw_line`,
- `CanvasItem::draw_polyline`,
- `CanvasItem::draw_arc`,
- `CanvasItem::draw_multiline`,
- `CanvasItem::draw_rect`,
- `CanvasItem::draw_circle`,
- `CanvasItem::draw_colored_polygon`,
- `RenderingServer::canvas_item_add_line`,
- `RenderingServer::canvas_item_add_rect`,
- `RenderingServer::canvas_item_add_circle`.


Resolves https://github.com/godotengine/godot-proposals/issues/7310.

---

Notes:
- In the code there were used `Color(1, 1, 1, 1)`, `Color(1, 1, 1)`, `Color(1.0, 1.0, 1.0)`. They are equivalent, I've changed all of them to `Color(1, 1, 1, 1)` which seemed to be used the most. Could be changed to some other variant if preferred. *Edit: it was done in `canvas_item.h/cpp`. But e.g. in `RenderingServer.h/cpp` `Color(1, 1, 1)` seems to be used instead, haven't touched these.*

kleonc committed:
Default to opaque white color in CanvasItem::draw_* methods

Calinou commented:
> I haven't added defaults for the `RenderingServer::canvas_item_add_*` methods.

I think we should do that as well for consistency.

clayjohn commented:
Looks like this will require compatibility methods for GDExtension and C#. Take a look at the "files changed" tab for more info

kleonc commented:
> > I haven't added defaults for the `RenderingServer::canvas_item_add_*` methods.
> 
> I think we should do that as well for consistency.

Changed some.

> Looks like this will require compatibility methods for GDExtension and C#. Take a look at the "files changed" tab for more info

Added compat methods, based on #76577. Kinda no idea what I'm doing though. :upside_down_face:

kleonc commented on servers/rendering_server.h:
```diff
@@ -70,6 +70,12 @@ class RenderingServer : public Object {
 	static void _bind_methods();
+#ifndef DISABLE_DEPRECATED
+	static void _bind_compatibility_methods();
+	void canvas_item_add_line_79666(RID p_item, const Point2 &p_from, const Point2 &p_to, const Color &p_color, float p_width = -1.0, bool p_antialiased = false);
+	void canvas_item_add_rect_79666(RID p_item, const Rect2 &p_rect, const Color &p_color);
+	void canvas_item_add_circle_79666(RID p_item, const Point2 &p_pos, float p_radius, const Color &p_color);
```
Note these are for pure virtual methods. Is it fine like this? :thinking:

raulsntos commented:
We currently don't generate compat methods for C#, so they have to be added manually:

```diff
diff --git a/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs b/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs
index 41cea031ab..b8f8ff7927 100644
--- a/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs
+++ b/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs
@@ -24,6 +24,65 @@ partial class AnimationNode
     }
 }
 
+partial class CanvasItem
+{
+    /// <inheritdoc cref="DrawArc(Vector2, float, float, float, int, Nullable{Color}, float, bool)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawArc(Vector2 center, float radius, float startAngle, float endAngle, int pointCount, Color color, float width, bool antialiased)
+    {
+        DrawArc(center, radius, startAngle, endAngle, pointCount, new Nullable<Color>(color), width, antialiased);
+    }
+
+    /// <inheritdoc cref="DrawCircle(Vector2, float, Nullable{Color})"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawCircle(Vector2 position, float radius, Color color)
+    {
+        DrawCircle(position, radius, new Nullable<Color>(color));
+    }
+
+    /// <inheritdoc cref="DrawColoredPolygon(Vector2[], Nullable{Color}, Vector2[], Texture2D)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawColoredPolygon(Vector2[] points, Color color, Vector2[] uvs, Texture2D texture)
+    {
+        DrawColoredPolygon(points, new Nullable<Color>(color), uvs, texture);
+    }
+
+    /// <inheritdoc cref="DrawDashedLine(Vector2, Vector2, Nullable{Color}, float, float, bool)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawDashedLine(Vector2 from, Vector2 to, Color color, float width, float dash, bool aligned)
+    {
+        DrawDashedLine(from, to, new Nullable<Color>(color), width, dash, aligned);
+    }
+
+    /// <inheritdoc cref="DrawLine(Vector2, Vector2, Nullable{Color}, float, bool)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawLine(Vector2 from, Vector2 to, Color color, float width, bool antialiased)
+    {
+        DrawLine(from, to, new Nullable<Color>(color), width, antialiased);
+    }
+
+    /// <inheritdoc cref="DrawMultiline(Vector2[], Nullable{Color}, float)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawMultiline(Vector2[] points, Color color, float width)
+    {
+        DrawMultiline(points, new Nullable<Color>(color), width);
+    }
+
+    /// <inheritdoc cref="DrawPolyline(Vector2[], Nullable{Color}, float, bool)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawPolyline(Vector2[] points, Color color, float width, bool antialiased)
+    {
+        DrawPolyline(points, new Nullable<Color>(color), width, antialiased);
+    }
+
+    /// <inheritdoc cref="DrawRect(Rect2, Nullable{Color}, bool, float)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void DrawRect(Rect2 rect, Color color, bool filled, float width)
+    {
+        DrawRect(rect, new Nullable<Color>(color), filled, width);
+    }
+}
+
 partial class CodeEdit
 {
     /// <inheritdoc cref="AddCodeCompletionOption(CodeCompletionKind, string, string, Nullable{Color}, Resource, Nullable{Variant})"/>
@@ -81,6 +140,30 @@ partial class RenderingDevice
     }
 }
 
+partial class RenderingServer
+{
+    /// <inheritdoc cref="CanvasItemAddCircle(Rid, Vector2, float, Nullable{Color})"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static void CanvasItemAddCircle(Rid item, Vector2 pos, float radius, Color color)
+    {
+        CanvasItemAddCircle(item, pos, radius, new Nullable<Color>(color));
+    }
+
+    /// <inheritdoc cref="CanvasItemAddLine(Rid, Vector2, Vector2, Nullable{Color}, float, bool)"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static void CanvasItemAddLine(Rid item, Vector2 from, Vector2 to, Color color, float width, bool antialiased)
+    {
+        CanvasItemAddLine(item, from, to, new Nullable<Color>(color), width, antialiased);
+    }
+
+    /// <inheritdoc cref="CanvasItemAddRect(Rid, Rect2, Nullable{Color})"/>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static void CanvasItemAddRect(Rid item, Rect2 rect, Color color)
+    {
+        CanvasItemAddRect(item, rect, new Nullable<Color>(color));
+    }
+}
+
 partial class RichTextLabel
 {
     /// <inheritdoc cref="PushList(int, ListType, bool, string)"/>

```

dsnopek commented:
> Added compat methods, based on #76577. Kinda no idea what I'm doing though.

Thanks! I've skimmed them and they look good to me. :-)

Unfortunately, they won't actually work until after PR https://github.com/godotengine/godot/pull/79702 is merged. I think it'd be good for that to be merged first, so that CI can double check that you've gotten them all correct.

kleonc commented:
> Unfortunately, they won't actually work until after PR #79702 is merged. I think it'd be good for that to be merged first, so that CI can double check that you've gotten them all correct.

Rebased, seeing #79702 was merged. Hopefully they are correct, will see.

Anyway, thanks all of you for the help with this! :slightly_smiling_face:

dsnopek commented:
> > Unfortunately, they won't actually work until after PR #79702 is merged. I think it'd be good for that to be merged first, so that CI can double check that you've gotten them all correct.
> 
> Rebased, seeing #79702 was merged. Hopefully they are correct, will see.

Unfortunately, CI is still not happy:

```
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_arc', from CFD4FBAB to 6031513E. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_colored_polygon', from 62E3D9E1 to B5F97BEA. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_dashed_line', from 81A7290C to EEC0FBE7. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_line', from 96057C42 to 89E57661. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_multiline', from FC2AB533 to 3BFB36E2. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_polyline', from F8E6DB22 to C0C2AA6D. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/CanvasItem/methods/draw_rect', from 0507B53D to B5FC4F88. This means that the function has changed and no compatibility function was provided.
```

Looking at the code, perhaps it's because all the orignal `DEFVAL()`'s aren't provided when registering the compatibility methods?

For example, for `draw_arc()` I think it maybe should be:

```
ClassDB::bind_compatibility_method(D_METHOD("draw_arc", "center", "radius", "start_angle", "end_angle", "point_count", "color", "width", "antialiased"), &CanvasItem::draw_arc_compat_79666, DEFVAL(-1.0), DEFVAL(false));
```

kleonc commented:
> Looking at the code, perhaps it's because all the orignal `DEFVAL()`'s aren't provided when registering the compatibility methods?

I was looking at this for some time and everything looked the same to me. Now that you've pointed it out it's so obvious, definitely needed another pair of eyes. Thanks again!

I wonder why it wasn't complaining about `RenderingServer::canvas_item_add_line` though, I've missed default values in compat binding for it too. :thinking:

Mickeon commented:
This is still desired, isn't it? Should it be redone? It _sounds_ more doable now. I don't recall experiencing trouble with C# compatibility bindings now.
